gcs/constraints/circuit/circuit.cc

```diff
@@ -0,0 +1,288 @@
+        proof.emit_proof_comment("Preventing sub-cycle for succ[" + to_string(prevent_idx->raw_value) + "] = " + to_string(prevent_value->raw_value));
+        proof_step << pos_var_data.at(prevent_idx->raw_value).plus_one_lines.at(prevent_value->raw_value).geq_line
+                   << " + ";
+    }
+    else {
+        proof.emit_proof_comment("Contradicting sub-cycle");
+    }
+
+    proof.emit_proof_line(proof_step.str(), ProofLevel::Current);
+}
+
+auto gcs::prevent_small_cycles(
+    const vector<IntegerVariableID> & succ,
+    const PosVarDataMap & pos_var_data,
+    const ConstraintStateHandle & unassigned_handle,
+    State & state) -> Inference
+{
+    auto result = Inference::NoChange;
+    auto & unassigned = any_cast<list<IntegerVariableID> &>(state.get_constraint_state(unassigned_handle));
+    auto n = succ.size();
+    auto end = vector<long>(n, -1);
+    auto known_ends = vector<long>{};
+    auto chain_lengths = vector<long>{};
+
+    for (auto var : unassigned) {
+        state.for_each_value_while(var, [&](Integer val) -> bool {
+            auto j0 = val.raw_value;
+            auto length = 0;
+            if (state.has_single_value(succ[j0]) && (end[j0] < 0)) {
+                auto j = j0;
+                do {
+                    j = state.optional_single_value(succ[j])->raw_value;
+                    length++;
+                    // Need to check this in case alldiff hasn't been fully propagated.
+                    if (j == j0) {
+                        state.infer_true(JustifyExplicitly{[&](Proof & proof) {
+                            output_cycle_to_proof(succ, j0, length, pos_var_data, state, proof);
+                        }});
+                        increase_inference_to(result, Inference::Contradiction);
+                        return false;
+                    }
+                } while (state.has_single_value(succ[j]));
+                end[j0] = j;
+                known_ends.emplace_back(j0);
+                chain_lengths.emplace_back(length);
+            }
+
+            return true;
+        });
+
+        if (result == Inference::Contradiction)
+            return result;
+    }
+
+    while (! known_ends.empty()) {
+        auto i = known_ends.back();
+        known_ends.pop_back();
+        auto length = chain_lengths.back();
+        chain_lengths.pop_back();
+        if (cmp_less(length, succ.size() - 1)) {
+            increase_inference_to(result,
+                state.infer(succ[end[i]] != Integer{i}, JustifyExplicitly{[&](Proof & proof) {
+                    output_cycle_to_proof(succ, i, length, pos_var_data, state, proof, Integer{end[i]}, Integer{i});
+                }}));
+        }
+        else {
+            increase_inference_to(result, state.infer(succ[end[i]] == Integer{i}, JustifyUsingRUP{}));
+        }
+    }
+    return result;
+}
+
+auto gcs::propagate_non_gac_alldifferent(const ConstraintStateHandle & unassigned_handle, State & state) -> innards::Inference
```

Comment:
Should we split the AllDifferent code up, and have both versions available as constraints?

Reply:
I can split these up, however, both prevent and non_gac_alldiff need to quickly access a list of unassigned variables which I currently have as a constraint state. If I make non_gac_alldiff it's own class "VCAllDifferent", both the VCAllDifferent constraint and the CircuitPrevent constraint are keeping track of unassigned variables separately which feels silly. Would it be reasonable to keep track of this in State itself and have a get_unassigned_variables method?

Reply:
Additionally I'm a little bit worried we won't be able to guarantee VCAllDifferent gets propagated before CircuitPrevent if it is a separate constraint, and we rely on this to stop infinite loops...

Reply:
For finding a constraint's unassigned variables quickly... This feels like it would be useful in a bunch of places. Potentially a constraint could also remove its triggers on certain variables during search. I think this might be a "solve it properly later" problem.

Reply:
> If you have a look on the `reasons` branch in `gcs/constraints/linear/propagate.hh`, there are functions that can be used to propagate a linear inequality without using a `Constraint` directly. Doing that for alldiff would let you call the propagation function at the start of your propagator, without needing to worry about it being a constraint that might not be called.

That would work, although since we also need the PB encoding would I duplicate the encoding logic both in Circuit and in VCAllDifferent? Or should the encoding function be available somewhere too?

Reply:
I don't think it would hurt to make the encoding function semi-public.